### ISSUE TITLE
vmagent: debug relabeling

### ISSUE THREAD

**Issue by jelmd** [VictoriaMetrics/VictoriaMetrics#1343]
To better understand, deep dive into relabeling it would help to log the metrics before and after the relabel stage and to prevent its submission to servers to avoid unnecessary poisoning.

**Comment by valyala** [comment]
I think it would be better adding a special mode to [vmctl](https://docs.victoriametrics.com/vmctl.html) tool, so it could accept a path to a file with relabeling rules, apply the relabeling rules the metrics in Prometheus exposition format from stdin and write the result to stdout. For example:

`cat metrics_before_relabeling.txt | vmctl relabel -relabelConfig=/path/to/relabel.yml > metrics_after_relabeling.txt`

The `metrics_before_relabeling.txt` file can be obtained directly from scrape target. For example:

`curl http://target-host:port/metrics > metrics_before_relabeling.txt`

@jelmd , @hagen1778 , what do you think about this proposal?

**Comment by jelmd** [comment]
Yes, sounds like a good idea, too. However, would be convenient, if one could use the promscrape config directly - not all endpoints are http://host:port/metrics (e.g. snmp_exporter). E.g. 
```
vmctl -f promscrape.yml -i metrics.txt [-o metrics-relabeled.txt]
```
If the input file does not yet exists, vmctl should scrape automagically the target and store the result incl. metadata in it (here `metrics.txt`). The full scrape configs are needed IMHO anyway, otherwise relabeling might not work as expected, because the "metadata" like `__*` are missing. Another advantage would be, that one can test its promcrape config in a safe way w/o disturbing anything.

Last but not least: Not sure, whether intention is to output the relabeled metrics, only. This would be not very helpful because metrics with labels are line-wise pretty long (80+ chars/line), so side-by-side comparison is pretty inconvenient (I guess, vmctl would also strip all comments and empty lines from the input, which makes it even harder). The other option would be diffing the input and output, but this usually leads to a result similar to  `cat input && cat output` - not very useful as well. Therefore I think, the output is only a real help, if it lists the full input (i.e. with __* labels) and right after the relabel result with __* for every single metric .

Sounds like a lot of code dups + new work wrt. metadata?

**Comment by valyala** [comment]
This sounds like a big amounts of work with many corner cases. Let's start with https://github.com/VictoriaMetrics/VictoriaMetrics/pull/1344 and then will decided whether `vmctl` needs the ability to apply relabeling.

**Comment by valyala** [comment]
Merged #1344 . Now it is possible to debug relabeling at the following stages:

* During service discovery with `relabel_debug: true` option in `scrape_config` section of `-promscrape.config` file.
* During metrics scraping with `metric_relabel_debug; true` option in `scrape_config` section of `-promscrape.config` file.
* Before writing the metrics to remote storage with `-remoteWrite.relabelDebug=true` and `-remoteWrite.urlRelabelDebug` command-line options.

See [these docs](https://docs.victoriametrics.com/vmagent.html#relabeling) for details.

**Comment by valyala** [comment]
The ability to debug relabeling at different stages of vmagent pipeline is available starting from [v1.61.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing the feature request as done.

**Comment by jelmd** [comment]
Thanx a lot! Just submitted PR #3412, so that one can also see/understand, why a metric got dropped.
